src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,755 @@
+		rc = ENOMEM;
+		goto out;
+	}
+
+	rc = dfs_local2global_all(hdl->dfs, &hdl->global);
+	if (rc) {
+		D_ERROR("Could not create global handler for dfs: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	rc = daos_eq_create(&hdl->eq);
+	if (rc) {
+		D_ERROR("Could not create event queue: %s (rc=%d)", d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
+		goto out;
+	}
+	hdl->eq_owner_pid = getpid();
+
+out:
+	if (rc) {
+		dfs_disconnect(hdl->dfs);
+
+		D_FREE(hdl->global.iov_buf);
+		D_FREE(hdl);
+		hdl = NULL;
+	}
+
+	PyList_SetItem(result, 0, PyLong_FromLong(rc));
+	PyList_SetItem(result, 1, PyLong_FromVoidPtr(hdl));
+
+	return result;
+}
+
+static PyObject *
+__shim_handle__torch_disconnect(PyObject *self, PyObject *args)
+{
+	struct dfs_handle *hdl = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "K", &hdl);
+
+	assert(hdl->dfs != NULL);
+
+	if (hdl->eq_owner_pid != getpid()) {
+		return PyLong_FromLong(EACCES);
+	}
+
+	int rc = dfs_disconnect(hdl->dfs);
+	if (rc) {
+		D_ERROR("Could not disconnect DFS: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+	rc = daos_eq_destroy(hdl->eq, DAOS_EQ_DESTROY_FORCE);
+	if (rc) {
+		D_ERROR("Could not destroy event queue: %s (rc=%d)", d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
+		goto out;
+	}
+
+	rc = dfs_fini();
+	if (rc) {
+		D_ERROR("Could not finalize DFS: %s (rc=%d)", strerror(rc), rc);
+		goto out;
+	}
+
+out:
+	D_FREE(hdl->global.iov_buf);
+	D_FREE(hdl);
+
+	return PyLong_FromLong(rc);
+}
+
+/*
+  DataLoader uses python multiprocessing with fork method to create its workers.
+  This function is the part of initialization routine for such workers, it reuses global connection
+  to the container and create an even queue for the worker process.
+ */
+static PyObject *
+__shim_handle__torch_worker_init(PyObject *self, PyObject *args)
+{
+	struct dfs_handle *hdl = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "K", &hdl);
+
+	if (hdl->eq_owner_pid == getpid()) {
+		return PyLong_FromLong(0);
+	}
+
+	int rc = dfs_init();
+	if (rc) {
+		return PyLong_FromLong(rc);
+	}
+
+	hdl->dfs          = NULL;
+	hdl->eq           = DAOS_HDL_INVAL;
+	hdl->eq_owner_pid = getpid();
+
+	rc = dfs_global2local_all(hdl->flags, hdl->global, &hdl->dfs);
+	if (rc) {
+		D_ERROR("Could not create local handler from global one: %s (rc=%d)", strerror(rc),
+			rc);
+		return PyLong_FromLong(rc);
+	}
+
+	rc = daos_eq_create(&hdl->eq);
+	if (rc) {
+		D_ERROR("Could not create event queue: %s (rc=%d)", d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__torch_recommended_dir_split(PyObject *self, PyObject *args)
+{
+	struct dfs_handle *hdl  = NULL;
+	char              *path = NULL;
+	dfs_obj_t         *obj  = NULL;
+	uint32_t           nr   = 0;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "Ls", &hdl, &path);
+
+	assert(hdl->dfs != NULL);
+
+	int rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		return PyLong_FromLong(-rc);
```

Comment:
hmm still see some neg errno returned here.

Reply:
The caller expects positive number of splits in case of success and a negative errno in case of a failure, I'm not sure what's wrong with this ?

Reply:
The caller expects positive number of splits in case of success and a negative errno in case of a failure, I'm not sure what's wrong with this ?

Reply:
ah ok. it's probably better then to return the error as the return value, but set a return variable for the slit to be consistent with the API. anyway not a big deal for me, and we can update that later.

Reply:
thanks, noted for upcoming patch!